cpu/native/socket_zep/socket_zep.c

```diff
@@ -259,6 +264,15 @@ static int _recv(netdev_t *netdev, void *buf, size_t len, void *info)
 
 static void _isr(netdev_t *netdev)
 {
+    socket_zep_t *dev = (socket_zep_t *)netdev;
+
+    /* discard packet when sleeping */
+    if (dev->disabled) {
+        uint8_t buffer[128];
+        real_read(dev->sock_fd, buffer, sizeof(buffer));
+        return;
+    }
```

Comment:
This could potentially take a long time and might break emulation behavior. How about [setting the receive buffer to `0` in sleep mode](https://stackoverflow.com/a/3097947) and re-setting it to its original size when leaving sleep mode?

Reply:
Might have the added benefit, that you won't need the `disabled` member and can just check the receive buffer length.

Reply:
I think I'll still need it to also prevent the device from sending.

Reply:
Hm I tried this patch

```patch
--- a/cpu/native/socket_zep/socket_zep.c
+++ b/cpu/native/socket_zep/socket_zep.c
@@ -264,15 +264,6 @@ static int _recv(netdev_t *netdev, void *buf, size_t len, void *info)
 
 static void _isr(netdev_t *netdev)
 {
-    socket_zep_t *dev = (socket_zep_t *)netdev;
-
-    /* discard packet when sleeping */
-    if (dev->disabled) {
-        uint8_t buffer[128];
-        real_read(dev->sock_fd, buffer, sizeof(buffer));
-        return;
-    }
-
     if (netdev->event_callback) {
         socket_zep_t *dev = (socket_zep_t *)netdev;
 
@@ -336,11 +327,17 @@ static int _get(netdev_t *netdev, netopt_t opt, void *value, size_t max_len)
 
 static int _set_state(socket_zep_t *dev, netopt_state_t state)
 {
+    int buf_size;
+
     switch (state) {
     case NETOPT_STATE_SLEEP:
+        buf_size = 0;
+        real_setsockopt(dev->sock_fd, SOL_SOCKET, SO_RCVBUF, &buf_size, sizeof(buf_size));
         dev->disabled = true;
         break;
     case NETOPT_STATE_IDLE:
+        buf_size = 128;
+        real_setsockopt(dev->sock_fd, SOL_SOCKET, SO_RCVBUF, &buf_size, sizeof(buf_size));
         dev->disabled = false;
         break;
     default:
```

and I would still receive UDP datagrams send with `udp send`.

Reply:
Alternatively you could close and re-open the socket as also proposed in the linked thread.